pkg/controller/nodelifecycle/scheduler/taint_manager_test.go

```diff
@@ -614,88 +684,36 @@ func TestUpdateNodeWithMultiplePods(t *testing.T) {
+		controller := NewNoExecuteTaintManager(ctx, fakeClientset, getPodFromClientset(fakeClientset), (&nodeHolder{node: item.newNode}).getNode, getPodsAssignedToNode(fakeClientset))
 		controller.recorder = testutil.NewFakeRecorder()
-		go controller.Run(context.TODO())
+		go controller.Run(ctx)
 		controller.NodeUpdated(item.oldNode, item.newNode)
 
-		startedAt := time.Now()
-		for i := range item.expectedDeleteTimes {
-			if i == 0 || item.expectedDeleteTimes[i-1].timestamp != item.expectedDeleteTimes[i].timestamp {
-				// compute a grace duration to give controller time to process updates. Choose big
-				// enough intervals in the test cases above to avoid flakes.
-				var increment time.Duration
-				if i == len(item.expectedDeleteTimes)-1 || item.expectedDeleteTimes[i+1].timestamp == item.expectedDeleteTimes[i].timestamp {
-					increment = 500 * time.Millisecond
-				} else {
-					increment = ((item.expectedDeleteTimes[i+1].timestamp - item.expectedDeleteTimes[i].timestamp) / time.Duration(2))
-				}
+		actions := getDeleteActionsWithCertainPods(fakeClientset, item.expectDelete)
 
-				sleepTime := item.expectedDeleteTimes[i].timestamp - time.Since(startedAt) + increment
-				if sleepTime < 0 {
-					sleepTime = 0
-				}
-				t.Logf("Sleeping for %v", sleepTime)
-				time.Sleep(sleepTime)
+		podDeleted := false
+		for _, action := range actions {
+			if action.GetVerb() == "delete" && action.GetResource().Resource == "pods" {
+				podDeleted = true
 			}
+		}
 
-			for delay, podName := range item.expectedDeleteTimes[i].names {
-				deleted := false
-				for _, action := range fakeClientset.Actions() {
-					deleteAction, ok := action.(clienttesting.DeleteActionImpl)
-					if !ok {
-						t.Logf("Found not-delete action with verb %v. Ignoring.", action.GetVerb())
-						continue
-					}
-					if deleteAction.GetResource().Resource != "pods" {
-						continue
-					}
-					if podName == deleteAction.GetName() {
-						deleted = true
-					}
-				}
-				if !deleted {
-					t.Errorf("Failed to deleted pod %v after %v", podName, delay)
-				}
-			}
-			for _, action := range fakeClientset.Actions() {
-				deleteAction, ok := action.(clienttesting.DeleteActionImpl)
-				if !ok {
-					t.Logf("Found not-delete action with verb %v. Ignoring.", action.GetVerb())
-					continue
-				}
-				if deleteAction.GetResource().Resource != "pods" {
-					continue
-				}
-				deletedPodName := deleteAction.GetName()
-				expected := false
-				for _, podName := range item.expectedDeleteTimes[i].names {
-					if podName == deletedPodName {
-						expected = true
-					}
-				}
-				if !expected {
-					t.Errorf("Pod %v was deleted even though it shouldn't have", deletedPodName)
-				}
-			}
-			fakeClientset.ClearActions()
+		if !podDeleted {
+			t.Errorf("%v: Unexpected test result. Expected delete true, got %v", item.description, podDeleted)
 		}
 
-		close(stopCh)
+		cancel()
```

Comment:
Call cancel in defer. This will make sure it is actually called.

Reply:
Fixed